parlai/agents/bart/modules.py

```diff
@@ -27,26 +26,37 @@ def output(self, tensor: torch.Tensor) -> torch.Tensor:
         output = F.linear(tensor, self.embeddings.weight)
         return output
 
-    def decode_forced(
-        self, encoder_states: Tuple[Any, ...], ys: torch.LongTensor
-    ) -> Tuple[torch.FloatTensor, torch.LongTensor]:
+    def _get_initial_forced_decoder_input(self, bsz: int, inputs: torch.LongTensor):
         """
-        Decode with a fixed, true sequence, computing loss.
+        Return initial input to the decoder.
+
+        Override TGA._get_initial_forced_decoder_input to seed EOS BOS.
 
-        Overriding `TGM.decode_forced` to bypass assertion that BOS is not present, and
-        additionally insert EOS as first token
+        :param bsz:
+            batchsize
+        :param inputs:
+            inputs to decode
+
+        :return initial_input:
+            initial input for the decoder.
         """
-        bsz = ys.size(0)
-        seqlen = ys.size(1)
-        inputs = ys.narrow(1, 0, seqlen - 1)
-        inputs = torch.cat(
-            [
-                torch.LongTensor([self.END_IDX]).detach().expand(bsz, 1).to(inputs),
-                inputs,
-            ],
-            1,
+        tens = (
+            torch.LongTensor([self.END_IDX, self.START_IDX])
+            .to(inputs)
+            .detach()
+            .expand(bsz, 2)
         )
-        latent, _ = self.decoder(inputs, encoder_states)
-        logits = self.output(latent)
-        _, preds = logits.max(dim=2)
-        return logits, preds
+        return torch.cat([tens, inputs], 1)
+
+    def reorder_decoder_incremental_state(
+        self,
+        incremental_state: Dict[str, Any],
+        inds: Union[List[int], torch.LongTensor],
+    ) -> Optional[Dict[str, Any]]:
+        """
+        Incremental state is weird to handle when we seed decoder with two inputs
+        initially.
+
+        We leave as a future exercise.
+        """
+        return None
```

Comment:
Hmm so I imagine generation with BERT will be a lot slower (but, er, correct) now?

Reply:
slower yes; a lot slower? maybe; probably not too significantly. if it becomes unwieldy in the future we can address then